core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4

```diff
@@ -841,6 +856,121 @@ pathSpecification
     : pathElement (',' pathElement)*
     ;
 
+functionSpecification
+    : FUNCTION qualifiedName parameterDeclarationList
+        returnsClause routineCharacteristic*
+        (returnStatement | compoundStatement)
+    ;
+
+parameterDeclarationList
+    : '(' (parameterDeclaration (',' parameterDeclaration)*)? ')'
+    ;
+
+parameterDeclaration
+    : identifier? type
+    ;
+
+returnsClause
+    : RETURNS type
+    ;
+
+routineCharacteristic
+    : LANGUAGE identifier               #languageCharacteristic
+    | NOT? DETERMINISTIC                #deterministicCharacteristic
+    | RETURNS NULL ON NULL INPUT        #returnsNullOnNullInputCharacteristic
+    | CALLED ON NULL INPUT              #calledOnNullInputCharacteristic
+    | SECURITY (DEFINER | INVOKER)      #securityCharacteristic
+    | COMMENT string                    #commentCharacteristic
+    ;
+
+routineStatement
+    : returnStatement
+    | assignmentStatement
+    | simpleCaseStatement
+    | searchedCaseStatement
+    | ifStatement
+    | iterateStatement
+    | leaveStatement
+    | compoundStatement
+    | loopStatement
+    | whileStatement
+    | repeatStatement
```

Comment:
You can inline all of these here and use alternative labels instead. The rules are referenced in a single place so have separate entries is not necessary and makes navigation a little harder.